src/pydicom/pixels/decoders/base.py

```diff
@@ -0,0 +1,2004 @@
+            * Encapsulated pixel data will be yielded as :class:`bytearray`.
+
+            8-bit pixel data encoded as **OW** using Explicit VR Big Endian will
+            be yielded as-is and may need byte-swapping. To facilitate this
+            an extra byte before the expected start (for and odd `index`) or after
+            the expected end (for an even `index`) is yielded if the frame contains
+            an odd number of pixels.
+        """
+        runner = DecodeRunner(self.UID)
+        runner.set_source(src)
+        runner.set_options(**kwargs)
+        runner.set_decoders(self._validate_decoders(decoding_plugin))
+
+        if validate:
+            runner.validate_options()
+            runner.validate_buffer()
+
+        if self.is_encapsulated and not indices:
+            yield from runner.iter_decode()
+            return
+
+        if self.is_native:
+            func = self._as_buffer_native
+        else:
+            func = self._as_buffer_encapsulated
+
+        indices = indices if indices else range(runner.number_of_frames)
+        for index in indices:
+            yield func(runner, index)
+
+    @property
+    def is_available(self) -> bool:
+        """Return ``True`` if the decoder has plugins available that can be
+        used to decode data, ``False`` otherwise.
+        """
+        # Decoders for public non-compressed syntaxes are always available
+        if self.is_native:
+            return True
+
+        return bool(self._available)
+
+    @property
+    def is_encapsulated(self) -> bool:
+        """Return ``True`` if the decoder is for an encapsulated transfer
+        syntax, ``False`` otherwise.
+        """
+        return self.UID.is_encapsulated
+
+    @property
+    def is_native(self) -> bool:
+        """Return ``True`` if the decoder is for an native transfer
+        syntax, ``False`` otherwise.
+        """
+        return not self.is_encapsulated
+
+    @property
+    def missing_dependencies(self) -> list[str]:
+        """Return nice strings for plugins with missing dependencies as
+        list[str].
+        """
+        s = []
+        for label, deps in self._unavailable.items():
+            if not deps:
+                # A plugin might have no dependencies and be unavailable for
+                #   other reasons
+                s.append(f"{label} - plugin indicating it is unavailable")
+            elif len(deps) > 1:
+                s.append(f"{label} - requires {', '.join(deps[:-1])} and {deps[-1]}")
+            else:
+                s.append(f"{label} - requires {deps[0]}")
+
+        return s
+
+    @property
+    def name(self) -> str:
+        """Return the name of the decoder as :class:`str`."""
+        return f"{self.UID.keyword}Decoder"
+
+    def remove_plugin(self, label: str) -> None:
+        """Remove a plugin from the decoder.
+
+        .. warning::
+
+            This method is not thread-safe.
+
```

Comment:
Mostly I'm concerned about there being a single instance of the specific UID decoder, so any changes to it in one thread  will propagate to the other threads. I've been trying to make everything threading/multiprocessing friendly (hopefully) since going off our citations in published papers a significant use case for our pixel data handling is machine learning/image processing pipelines. 

I think this affects the encoder implementation too, I'm planning on refactoring it when I add in support for encoding J2K.